### ISSUE TITLE
Fails to parse conformance tests (ValidationException: Failed to match union type)

### ISSUE THREAD

**Issue by jdidion** [common-workflow-language/cwljava#29]
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/stage_file_array_basename_and_entryname.cwl
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/docker-array-secondaryfiles.cwl
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/optional-output.cwl
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/record-in-secondaryFiles.cwl
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/initialworkdirrequirement-docker-out.cwl
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/secondaryfiles/rename-outputs.cwl
   This is a cwlVersion: v1.0 file, so it shouldn't parse, when upgraded to v1.2 it does parse with cwljava
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/stage_file_array.cwl
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/stage_file_array_basename.cwl
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/stage_array.cwl
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/initialworkdir-glob-fullpath.cwl
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/record-out-secondaryFiles.cwl
- [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/secondaryfiles/rename-inputs.yml
  This is a cwlVersion: v1.0 file, so it shouldn't parse, but even when upgraded to v1.2 it still doesn't parse

* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/timelimit.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/timelimit2.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/timelimit3.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/cores_float.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/mixed-versions/tool-v12.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/iwd/iwdr_dir_literal_real_file.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/string-interpolation/js-quote.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/iwd/iwd-passthrough4.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/stage-array-dirs.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/dir5.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/storage_float.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/dynresreq.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/dynresreq-dir.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/dynresreq-default.cwl
* [x] https://github.com/common-workflow-language/cwl-v1.2/blob/main/tests/linkfile.cwl

```
org.w3id.cwl.cwl1_2.utils.ValidationException: Failed to match union type
[info]   Trying 'RecordField'
[info]     the `inputs` field is not valid because:
[info]       Failed to match union type
[info]         Expected object with Java type of java.util.List but got java.util.LinkedHashMap
[info]         Trying 'RecordField'
[info]           the `secondaryFiles` field is not valid because:
[info]             Failed to match union type
[info]               Expected null
[info]               Expected object with Java type of java.util.Map but got java.util.ArrayList
[info]               Failed to match union type
[info]                 Expected object with Java type of java.util.List but got java.lang.String
[info]                 Expected object with Java type of java.util.Map but got java.lang.String
[info]   Trying 'RecordField'
[info]     the `inputs` field is not valid because:
[info]       Failed to match union type
[info]         Expected object with Java type of java.util.List but got java.util.LinkedHashMap
[info]         Trying 'RecordField'
[info]           the `secondaryFiles` field is not valid because:
[info]             Failed to match union type
[info]               Expected null
[info]               Expected object with Java type of java.util.Map but got java.util.ArrayList
[info]               Failed to match union type
[info]                 Expected object with Java type of java.util.List but got java.lang.String
[info]                 Expected object with Java type of java.util.Map but got java.lang.String
[info]     the `expression` field is not valid because:
[info]       Expected object with Java type of java.lang.String but got null
[info]   Trying 'RecordField'
[info]     the `inputs` field is not valid because:
[info]       Failed to match union type
[info]         Expected object with Java type of java.util.List but got java.util.LinkedHashMap
[info]         Trying 'RecordField'
[info]           the `secondaryFiles` field is not valid because:
[info]             Failed to match union type
[info]               Expected null
[info]               Expected object with Java type of java.util.Map but got java.util.ArrayList
[info]               Failed to match union type
[info]                 Expected object with Java type of java.util.List but got java.lang.String
[info]                 Expected object with Java type of java.util.Map but got java.lang.String
[info]     the `steps` field is not valid because:
[info]       Expected object with Java type of java.util.List but got null
[info]   Trying 'RecordField'
[info]     the `inputs` field is not valid because:
[info]       Failed to match union type
[info]         Expected object with Java type of java.util.List but got java.util.LinkedHashMap
[info]         Trying 'RecordField'
[info]           the `secondaryFiles` field is not valid because:
[info]             Failed to match union type
[info]               Expected null
[info]               Expected object with Java type of java.util.Map but got java.util.ArrayList
[info]               Failed to match union type
[info]                 Expected object with Java type of java.util.List but got java.lang.String
[info]                 Expected object with Java type of java.util.Map but got java.lang.String
[info]   Expected object with Java type of java.util.List but got java.util.LinkedHashMap
[info]   at org.w3id.cwl.cwl1_2.utils.UnionLoader.load(UnionLoader.java:31)
[info]   at org.w3id.cwl.cwl1_2.utils.Loader.documentLoad(Loader.java:41)
[info]   at org.w3id.cwl.cwl1_2.utils.RootLoader.loadDocument(RootLoader.java:20)
[info]   at org.w3id.cwl.cwl1_2.utils.RootLoader.loadDocument(RootLoader.java:86)
[info]   at org.w3id.cwl.cwl1_2.utils.RootLoader.loadDocument(RootLoader.java:45)
```

**Comment by mr-c** [comment]
These are all due to https://github.com/common-workflow-language/schema_salad/issues/349 ( lack of support in the codegen classes for the new way we represent secondaryFiles in the CWL v1.1+ schema)

**Comment by mr-c** [comment]
Some of these were fixed in #32 and confirmed in #34 , so I've marked them as such

**Comment by jdidion** [comment]
I can confirm those tests were fixed, but it also looks like some that were previously passing are now broken:

[list moved to original post]

**Comment by jdidion** [comment]
all fixed now by #38
